### DISCUSSION TITLE
Configuring RVPipelines Implementation of get_config() or get_configs()

### DISCUSSION THREAD

**Discussion by ammarsani-my** [azavea/raster-vision#1518] [Q&A] [answered]
Hi. In your documentation here: https://docs.rastervision.io/en/latest/pipelines.html#configuring-rvpipelines

> Each (subclass of) [RVPipeline](https://docs.rastervision.io/en/latest/api_reference/_generated/rastervision.core.rv_pipeline.rv_pipeline.RVPipeline.html#rastervision.core.rv_pipeline.rv_pipeline.RVPipeline) is configured by returning an instance of (a subclass of) [RVPipelineConfigs](https://docs.rastervision.io/en/latest/api_reference/_generated/rastervision.core.rv_pipeline.rv_pipeline_config.RVPipelineConfig.html#rastervision.core.rv_pipeline.rv_pipeline_config.RVPipelineConfig) from a get_config() function in a Python module. It’s also possible to return a list of [RVPipelineConfigs](https://docs.rastervision.io/en/latest/api_reference/_generated/rastervision.core.rv_pipeline.rv_pipeline_config.RVPipelineConfig.html#rastervision.core.rv_pipeline.rv_pipeline_config.RVPipelineConfig) from get_configs(), which will be executed in parallel.

I could not find the implementation of get_config() or get configs() in any of classes implementing RVPipeline as mentioned there. May I know what is that means?

Thanks!

**Comment by AdeelH** [comment]
`get_config()` is defined by the user and returns a pipeline config. See, for example, https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py#L33 .

When you run Raster Vision from the command line, it runs this function to get the pipeline config and then proceeds to run the pipeline.

**Reply by ammarsani-my** [reply]
Thanks for the response @AdeelH. Preparing a pipeline file as in reference you share here was exactly what we were successfully did here. We're so grateful with what RV could offers up to this point.

However when we rereading your documentation this time, what we do understand from the quoted statement above is any class derived from RVPipeline such as SegmentationPipeline and the others were expected to implement both methods, `get_config()` and `get_configs()` in each of this derived class.

But what is supposed to say that these derived classes is expecting for a pipeline .py file that has `get_config()` or `get_configs()` in it.

Am I correct?

**Comment by AdeelH** [comment, marked as answer]
>Preparing a pipeline file as in reference you share here was exactly what we were successfully did here

In that case, you are using it correctly and don't need to change anything.

>what we do understand from the quoted statement above is any class derived from RVPipeline such as SegmentationPipeline and the others were expected to implement both methods, get_config() and get_configs() in each of this derived class

This is incorrect. These classes do not have to implement these.

>that these derived classes is expecting for a pipeline .py file that has get_config() or get_configs() in it.

Mostly correct. The derived classes don't care about `get_config()` but rather the pipeline config that it returns. `get_config()` is [executed in rastervision.pipeline.cli](https://github.com/azavea/raster-vision/blob/master/rastervision_pipeline/rastervision/pipeline/cli.py#L74).

**Reply by ammarsani-my** [reply]
Okay got it. Thanks @AdeelH
